sw/device/lib/crypto/impl/status_unittest.cc

```diff
@@ -0,0 +1,128 @@
+}
+
+TEST(Status, ErrorMacrosNotOk) {
+  // Error macros should evaluate to non-OK statuses.
+  EXPECT_EQ(status_ok(OTCRYPTO_BAD_ARGS), false);
+  EXPECT_EQ(status_ok(OTCRYPTO_RECOV_ERR), false);
+  EXPECT_EQ(status_ok(OTCRYPTO_FATAL_ERR), false);
+  EXPECT_EQ(status_ok(OTCRYPTO_ASYNC_INCOMPLETE), false);
+}
+
+TEST(Status, InterpretErrorMacros) {
+  // Error macros should translate to the crypto status implied by their name.
+  EXPECT_EQ(status_interpret(OTCRYPTO_OK), kCryptoStatusOK);
+  EXPECT_EQ(status_interpret(OTCRYPTO_BAD_ARGS), kCryptoStatusBadArgs);
+  EXPECT_EQ(status_interpret(OTCRYPTO_RECOV_ERR), kCryptoStatusInternalError);
+  EXPECT_EQ(status_interpret(OTCRYPTO_FATAL_ERR), kCryptoStatusFatalError);
+  EXPECT_EQ(status_interpret(OTCRYPTO_ASYNC_INCOMPLETE),
+            kCryptoStatusAsyncIncomplete);
+}
+
+static __attribute__((noinline)) crypto_status_t try_interpret(
+    status_t status) {
+  HARDENED_TRY_INTERPRET(status);
+  return kCryptoStatusOK;
+}
+
+TEST(Status, TryInterpretOk) {
+  // Hardened OK should result in an OK status.
+  EXPECT_EQ(try_interpret(OTCRYPTO_OK), kCryptoStatusOK);
+}
+
+TEST(Status, TryInterpretNonHardenedOk) {
+  // Non-hardened OK should result in an error.
+  EXPECT_EQ(try_interpret(OK_STATUS()), kCryptoStatusInternalError);
+}
+
+TEST(Status, TryInterpretErrors) {
+  // Error macros should result in error statuses.
+  EXPECT_EQ(try_interpret(OTCRYPTO_BAD_ARGS), kCryptoStatusBadArgs);
+  EXPECT_EQ(try_interpret(OTCRYPTO_RECOV_ERR), kCryptoStatusInternalError);
+  EXPECT_EQ(try_interpret(OTCRYPTO_FATAL_ERR), kCryptoStatusFatalError);
+  EXPECT_EQ(try_interpret(OTCRYPTO_ASYNC_INCOMPLETE),
+            kCryptoStatusAsyncIncomplete);
+}
+
+constexpr char test_mod_id[3] = {'X', 'Y', 'Z'};
+DECLARE_MODULE_ID(test_mod_id[0], test_mod_id[1], test_mod_id[2]);
+
+TEST(Status, ExtractStatusFieldsBadArgs) {
+  const char *code;
+  char mod_id[3];
+  int32_t line;
+  const char expected_code[] = "InvalidArgument";
+  int32_t expected_line = __LINE__ + 1;
+  EXPECT_EQ(status_extract(OTCRYPTO_BAD_ARGS, &code, &line, mod_id), true);
+
+  // Check the fields to ensure that the format of cryptolib errors matches the
+  // error format from the main status library.
+  EXPECT_EQ(memcmp(code, expected_code, ARRAYSIZE(expected_code)), 0);
+  EXPECT_THAT(mod_id, testing::ElementsAreArray(test_mod_id));
+  EXPECT_EQ(line, expected_line);
+}
+
+TEST(Status, ExtractStatusFieldsRecovErr) {
+  const char *code;
+  char mod_id[3];
+  int32_t line;
+  const char expected_code[] = "Aborted";
+  int32_t expected_line = __LINE__ + 1;
+  EXPECT_EQ(status_extract(OTCRYPTO_RECOV_ERR, &code, &line, mod_id), true);
+
+  // Check the fields to ensure that the format of cryptolib errors matches the
+  // error format from the main status library.
+  EXPECT_EQ(memcmp(code, expected_code, ARRAYSIZE(expected_code)), 0);
+  EXPECT_THAT(mod_id, testing::ElementsAreArray(test_mod_id));
+  EXPECT_EQ(line, expected_line);
```

Comment:
Super optional: Consider rolling these `ExtractStatus*` tests into a loop with a `SCOPED_TRACE` to tag failing cases.

```c++
TEST(Status, ExtractStatusFields) {
  struct TestCase {
    status_t status;
    std::string expected_code;
    int32_t expected_line;
  };

  constexpr TestCase kTestCases[] = {
    {.status = OTCRYPTO_BAD_ARGS, .expected_code = "InvalidArgument", .expected_line = __LINE__},
    // ...
  };

  for (const auto& test : kTestCases) {
    SCOPED_TRACE("expected_code: " + expected_code);

    const char *code = nullptr;
    char mod_id[3] = {0};
    int32_t line = 0;
    EXPECT_EQ(status_extract(OTCRYPTO_RECOV_ERR, &code, &line, mod_id), true);
  
    // Check the fields to ensure that the format of cryptolib errors matches the
    // error format from the main status library.
    EXPECT_EQ(std::string(code), test.expected_code);
    EXPECT_THAT(mod_id, testing::ElementsAreArray(kTestModId));
    EXPECT_EQ(line, test.expected_line);
  }
}
```